### ISSUE TITLE
x/build: record and collect compiler benchmarks

### ISSUE THREAD

**Issue by mdempsky** [golang/go#17167]
Recently we added a -bench flag to cmd/compile to benchmark compiler phases.  It might be interesting if the builders used this flag and then persisted the output file so we could monitor compiler performance over time.

There's a risk that measuring the standard library compile time is confounded with changes to the standard library, but I think only a couple packages (runtime, net, net/http) still seem under active development.  So it's probably still a decent signal.

/cc @bradfitz @griesemer

**Comment by bradfitz** [comment]
I've been logging more and more to Datastore & BigQuery, and it's on my short-term list to add more timing info there, as part of optimizing build+test speeds for #17104

So, this isn't much more work.

/cc @quentinmit

**Comment by quentinmit** [comment]
Yep, and I'm working on collecting benchmark results. This sounds great. Where does -bench put its output?

Also, which compile command would you instrument? There are a bunch involved in all.bash.

**Comment by mdempsky** [comment]
You can use -bench=foo.txt and the compiler will append to foo.txt, so the entire build can be accumulated into a single file.  The individual measurements indicate the package being compiled.

I would probably tweak all.bash so that the "go install std cmd" step adds -gcflags="-bench=/tmp/bench.txt" on the builders, and then save /tmp/bench.txt somewhere we can easily collect and analyze.

I think benchmarking the Go1 benchmarks' compile times would be nice too, since they're a fixed target.

**Comment by bradfitz** [comment]
We don't use all.bash on the builders. And make.bash already respects the env var $GO_GCFLAGS

**Comment by mdempsky** [comment]
I meant "all.bash" generically to refer to whatever build process was used by the builders.  GO_GCFLAGS sounds good to me if that means we can do it non-intrusively.

**Comment by davecheney** [comment]
SGTM!

**Comment by prattmic** [comment]
Yes, we have benchmark tracking now. I think the action item here would be to add some compiler phase benchmarks to x/benchmarks. Perhaps alongside our existing overall build benchmarks in https://cs.opensource.google/go/x/benchmarks/+/master:sweet/harnesses/go-build.go?
